File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/C._Золотая_лихорадка/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>


void solution() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(1)
    */
    long long m, n;
    std::cin >> m >> n;
    std::vector<std::pair<long long, long long>> coins;
    for (int i = 0; i < n; ++i) {
        long long cost, freq;
        std::cin >> cost >> freq;
        coins.push_back({cost, freq});
    }
    std::sort(coins.begin(), coins.end());
    long long profit = 0;
    for (int idx = n - 1; idx >= 0; --idx) {
        std::pair<long long, long long>& p = coins[idx];
        int t = std::min(m, p.second);
        profit += p.first * t;
        m -= t;
        if (!m) break;
    }
    std::cout << profit << "\n";
}


int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/C._Золотая_лихорадка/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(1)
    """
    m: int = int(sys.stdin.readline().rstrip())
    n: int = int(sys.stdin.readline().rstrip())
    coins: list[tuple[int, int]] = []
    for _ in range(n):
        cost, freq = map(int, sys.stdin.readline().rstrip().split())
        coins.append((cost, freq))
    coins.sort(reverse=True)
    profit: int = 0
    for cost, freq in coins:
        t: int = min(m, freq)
        profit += cost * t
        m -= t
        if not m: break
    sys.stdout.write(f'{profit}\n')


if __name__ == '__main__':
    solution()